iceoryx_hoofs/posix/time/source/deadline_timer.cpp

```diff
@@ -48,11 +51,21 @@ iox::units::Duration deadline_timer::remainingTime() const noexcept
     {
         return m_endTime - currentTime;
     }
-    return iox::units::Duration(std::chrono::milliseconds(0));
+    using namespace iox::units::duration_literals;
+    return 0_s;
 }
 
 iox::units::Duration deadline_timer::getCurrentMonotonicTime() noexcept
 {
-    return iox::units::Duration{std::chrono::steady_clock::now().time_since_epoch()};
+    timespec time_since_epoch{0, 0};
+    if (posix::posixCall(clock_gettime)(CLOCK_REALTIME, &time_since_epoch)
+            .failureReturnValue(-1)
+            .evaluate()
+            .has_error())
+    {
+        using namespace iox::units::duration_literals;
+        return 0_s;
```

Comment:
I'm hesitant to silently fail. Can you check under which conditions `clock_gettime` can fail?

My tendency is to call the error handler since we do not now what the user does after the call to this method. It might lead to a failure on the user side. Mid term we should add a `tryGetCurrentMonotonicTime` and encourage that use instead of something that can terminate.

Reply:
I would argue that this doesn't really matter because [`time_since_epoch`](https://en.cppreference.com/w/cpp/chrono/time_point/time_since_epoch) has the same behaviour. The caller needs to deal with the result of `0`.

Reply:
I just had a look at `clock_gettime` and it seems there are three possible errors
- EFAULT: tp points outside the accessible address space. 
- EPERM: clock_settime() does not have permission to set the clock indicated. 
- EINVAL: The clk_id specified is not supported on this system. 

The first one cannot happen since `tp` is on the stack. The second one also cannot happen since we are not using `clock_settime`. The last one might happen when someone ports iceoryx to a esoteric platform or someone creates a faulty image for a supported platform. In this case the class is completely broken and will never work. Therefore it makes sense to call the error handler here and fail early. In another PR an `IOX_UNREACHABLE` macro is in the works and that one would be ideal for this case since it will be unreachable in a properly configured system. In the meantime, could you add an `Ensures` or at least a todo to add the `IOX_UNREACHABLE`?

Reply:
I think that's very esoteric and nothing can go really wrong here, because a properly constructed `Duration` is returned.

If this is changed I would propose to change it consistently at all locations:

https://github.com/eclipse-iceoryx/iceoryx/blob/619514f1e98088f978935e5ec1601649024d2fdc/iceoryx_hoofs/time/source/duration.cpp#L55-L58

https://github.com/eclipse-iceoryx/iceoryx/blob/619514f1e98088f978935e5ec1601649024d2fdc/iceoryx_hoofs/source/log/building_blocks/console_logger.cpp#L58-L62

Is it ok for you if we take this to a follow-up PR?

Reply:
Being esoteric is exactly the reason I'd like it to have an early failure. I'd rather have a bug report with a `I get this error code` than `I get a zero Duration`. The former tells me exactly what happens. The latter would be either `clock_gettime` returning 0 or an error.

I don't mind if you change it for the logger but there the situation is a bit different. Having a timestamp is not the crucial part of the functionality and the logger would have some value even if `clock_gettime` is broken. `getCurrentMonotonicTime` only has one functionality and when `clock_gettime` is broken all of the functionality is broken.

Therefore I'd like to have the fast failure for `getCurrentMonotonicTime`. If you also want to have it in `createLogMessageHeader` feel free to change it.

Reply:
@mossmaurice if you have to commit anyway please either add a TODO or an `Ensures` and a `IOX_LOG(FATAL)` ... oh, in this case it might be a good idea not to have an `Ensures` in the `createLogMessageHeader`. I just remember why I needed the resilience :sweat_smile:

Reply:
... oh, another reason why an `Ensures` would be a good idea is that one could have code like this
```cpp
deadline_timer t{1_ms};
while(!t.has_expired()) {
    // polling code
}
```

One wouldn't even easily notice that the code is broken but just notice a high CPU load.

Reply:
@elBoberido Makes sense, added. I chose to wrap the call in a `EnsuresWithMsg` as this our strategy for non-recoverable errors (see #1902). I left the logger untouched and added a further comment.